include/nil/crypto3/zk/snark/arithmetization/plonk/detail/lookup_table_definition.hpp

```diff
@@ -0,0 +1,166 @@
+
+                        lookup_table_definition(const std::string table_name){
+                            this->table_name = table_name;
+                        }                        
+
+                        virtual void generate() = 0;
+
+                        const std::vector<std::vector<typename FieldType::value_type>> &get_table(){
+                            if(_table.size() == 0){
+                                generate();
+                            }
+                            return _table;
+                        }
+                        virtual ~lookup_table_definition() {};
+                    };
+
+                    template<typename FieldType>
+                    class filled_lookup_table_definition:public lookup_table_definition<FieldType>{
+                    public:
+                        filled_lookup_table_definition(std::shared_ptr<lookup_table_definition<FieldType>> &other):lookup_table_definition<FieldType>(other->table_name){
+                            this->_table = other->get_table();
+                        }
+                        virtual void generate() {};
+                        virtual ~filled_lookup_table_definition() {};
+                    };
+
+                    // Returned value -- new usable_rows.
+                    // All tables are necessary for circuit generation.
+                    template<typename FieldType, typename ArithmetizationParams>
+                    std::size_t pack_lookup_tables(
+                        const std::map<std::string, std::size_t> &lookup_table_ids,
+                        const std::map<std::string, std::shared_ptr<lookup_table_definition<FieldType>>> &lookup_tables, 
+                        plonk_constraint_system<FieldType, ArithmetizationParams> &bp,
+                        plonk_assignment_table<FieldType, ArithmetizationParams> &assignment,
+                        const std::vector<std::size_t> &constant_columns_ids,
+                        std::size_t usable_rows
+                    ){
+                        std::size_t usable_rows_after = usable_rows;
+
+                        // Compute first selector index.
+                        std::size_t cur_selector_id = 0;
+                        for(const auto &gate: bp.gates()){
+                            cur_selector_id = std::max(cur_selector_id, gate.selector_index);
+                        }
+                        for(const auto &lookup_gate: bp.lookup_gates()){
+                            cur_selector_id = std::max(cur_selector_id, lookup_gate.tag_index);
+                        }
+                        cur_selector_id++;
+
+                        // Allocate constant columns
+                        std::vector<plonk_column<FieldType>> constant_columns(constant_columns_ids.size());
+                        std::vector<plonk_column<FieldType>> selector_columns;
+                        for(std::size_t i = 0; i < constant_columns.size(); i++){

```

Comment:
Let's use constant_columns_ids.size(), to keep it consistent. 2 lines up we used it, when creating the vector.

Also it's probably faster this way:

std::vector<plonk_column<FieldType>> constant_columns(constant_columns_ids.size(), plonk_column<FieldType>(usable_rows, FieldType::value_type::zero()));